### PR TITLE
Color mode implementation

### PR DESCRIPTION
**Pull request by oxan**
API changes for esphome/esphome#2012.

### CODE REVIEW

File: aioesphomeapi/model.py [hunk 1]
```diff
@@ -207,26 +207,46 @@ class FanState(EntityState):
...
-    supports_white_value: bool = False
-    supports_color_temperature: bool = False
+    supported_color_modes: List[LightColorMode] = converter_field(
+        default_factory=list, converter=list
+    )
     min_mireds: float = 0.0
     max_mireds: float = 0.0
     effects: List[str] = converter_field(default_factory=list, converter=list)
 
+    # deprecated, do not use
```
Comment (OttoWinter) [inline aioesphomeapi/model.py:231]: Let's call these `legacy_...`.

Then let's add a `supported_color_modes_compat` method that assembles the `color_mode` values given the peer API version (like with climate presets). Same for the call in `client.py`.

Advantage then is that we have all the compat logic in aioesphomeapi, and not in HA.

File: aioesphomeapi/model.py [hunk 2]
```diff
@@ -207,26 +207,46 @@ class FanState(EntityState):
...
+
+
 @dataclass(frozen=True)
 class LightInfo(EntityInfo):
     supports_brightness: bool = False
-    supports_rgb: bool = False
-    supports_white_value: bool = False
-    supports_color_temperature: bool = False
+    supported_color_modes: List[LightColorMode] = converter_field(
+        default_factory=list, converter=list
```
Comment (OttoWinter) [inline aioesphomeapi/model.py:225]:
```suggestion
        default_factory=list, converter=LightColorMode.convert_list
```

To convert the values from the protobuf enums to our python enums

File: aioesphomeapi/model.py (hunk 1 above)
Comment (OttoWinter) [reply aioesphomeapi/model.py:231]: Ok scratch that about `client.py` - the client class doesn't have enough info to transform those calls. But for `LightInfo` we can do it:

```python3
    legacy_supports_rgb: bool = False
    legacy_supports_white_value: bool = False
    legacy_supports_color_temperature: bool = False

    def supported_color_modes_compat(self, api_version: APIVersion) -> List[LightColorMode]:
        if api_version < APIVersion(1, 6):
            key = (
                self.legacy_supports_rgb, 
                self.legacy_supports_white_value, 
                self.legacy_supports_color_temperature
            )
            # map legacy flags to color modes, 
            # key: (rgb, white, color_temp)
            modes_map = {
                (False, False, False): [],
                (False, False, True): [LightColorMode.COLOR_TEMPERATURE],
                (False, True, False): [],  # invalid
                (False, True, True): [],  # invalid
                (True, False, False): [LightColorMode.RGB],
                (True, False, True): [],  # invalid
                (True, True, False): [LightColorMode.RGB_WHITE],
                (True, True, True): [LightColorMode.RGB_COLOR_TEMPERATURE]
            }
            return modes_map[key]

        return self.supported_color_modes
```

Comment (oxan) [reply aioesphomeapi/model.py:231]: > Let's call these `legacy_...`.

Just to be sure, we don't care about backward compatibility here?

Comment (OttoWinter) [reply aioesphomeapi/model.py:231]: Not really, the package is only really used by HA afaik (and some smaller things, but none that would touch color modes I think). We can just create a major version bump.

**Review by OttoWinter — APPROVED** [review]

File: aioesphomeapi/model.py [hunk 3]
```diff
@@ -207,26 +207,73 @@ class FanState(EntityState):
...
+            return modes_map[key] if key in modes_map else []
+
+        return self.supported_color_modes
+
 
 @dataclass(frozen=True)
 class LightState(EntityState):
     state: bool = False
     brightness: float = 0.0
+    color_mode: LightColorMode = LightColorMode.UNKNOWN
```
Comment (OttoWinter) [inline aioesphomeapi/model.py:268]:
```suggestion
    color_mode: Optional[LightColorMode] = converter_field(
        default=LightColorMode.UNKNOWN, converter=LightColorMode.convert
    )
```

**Pull request merged**